Chatkit/Subscriptions/SubscriptionManager.swift

```diff
@@ -0,0 +1,79 @@
+
+protocol HasSubscriptionManager {
+    var subscriptionManager: SubscriptionManager { get }
+}
+
+protocol SubscriptionManager {
+    func subscribe(toType subscriptionType: SubscriptionType, sender: AnyObject, completion: @escaping SubscribeHandler)
+    func unsubscribe(fromType subscriptionType: SubscriptionType, sender: AnyObject)
+    func unsubscribeFromAll()
+}
+
+class ConcreteSubscriptionManager: SubscriptionManager {
+    
+    typealias Dependencies = HasSubscriptionFactory
+    typealias SubscriptionDetails = (subscription: Subscription, subscribers: NSHashTable<AnyObject>)
+    
+    private let dependencies: Dependencies
+    private var subscriptionDetailsByType: [SubscriptionType: SubscriptionDetails] = [:]
+    
+    init(dependencies: Dependencies) {
+        self.dependencies = dependencies
+    }
+    
+    // MARK: SubscriptionManager
+    
+    func subscribe(toType subscriptionType: SubscriptionType, sender: AnyObject, completion: @escaping SubscribeHandler) {
+        
+        let subscriptionDetails: SubscriptionDetails
+        
+        // Check if we are already have SubscriptionDetails for the specified SubscriptionType and if so use that
+        if let subscriptionDetailsTmp = subscriptionDetailsByType[subscriptionType] {
+            subscriptionDetails = subscriptionDetailsTmp
+        } else {
+            let subscription = dependencies.subscriptionFactory.makeSubscription(subscriptionType: subscriptionType)
+            let subscribers = NSHashTable<AnyObject>.weakObjects()
+            subscriptionDetails = (subscription: subscription, subscribers: subscribers)
+            subscriptionDetailsByType[subscriptionType] = subscriptionDetails
+        }
+
+        // Hold a *weak* reference to the sender as a "subscriber"
+        subscriptionDetails.subscribers.add(sender)
+        
+        // TODO: Subscription unhappy paths
```

Comment:
I was not able to find a LeanKit card for this work. Should we create one?

Reply:
I have created card "Subscription Unhappy Paths" 
https://pusherpm.leankit.com/board/804539501